### ISSUE TITLE
Evaluation Joints

### ISSUE THREAD

**Issue by Tim-Salzmann** [facebookresearch/QuaterNet#10]
Hi,

first of all: I found your paper very inspiring! Keep up the good work!

Looking at some state of the art algorithms I have seen that it is quite common to ignore the first joint. As far as I understood this is done to ignore the absolute rotation of the body (e.g. pelvic bone) with respect to the MOCAP origin?!

However I feel like you are ignoring the first and the second joint for evaluation.

The first joint is ignored during dataset preparation:
https://github.com/facebookresearch/QuaterNet/blob/9d8485b732b0a44b99b6cf4b12d3915703507ddc/prepare_data_short_term.py#L72

However, from the resulting dataset the first joint is again ignored during evaluation [:, 3:]:
https://github.com/facebookresearch/QuaterNet/blob/9d8485b732b0a44b99b6cf4b12d3915703507ddc/test_short_term.py#L82

Could you please clarify why this is done and if this is in line with the other state of the art algorithms you are comparing to?

Thanks

**Comment by dariopavllo** [comment]
Our evaluation methodology follows Martinez et al. / RNNexp. You can check the following scripts to compare our implementation with theirs:

Model evaluation:
https://github.com/una-dinosauria/human-motion-prediction/blob/b8f0cb57846e9d2d5688a3e30ab7d8840bb65f38/src/translate.py#L212-L221

Baselines evaluation:
https://github.com/una-dinosauria/human-motion-prediction/blob/b8f0cb57846e9d2d5688a3e30ab7d8840bb65f38/src/baselines.py#L71-L72

In the preprocessed dataset they provide, the first 3 floats represent the global translation of the root joint (but this is corrupted and therefore useless), and the next triplet of floats represents the rotation of the root joint. The evaluation discards the root joint, which is the common practice in this task.

At some point I think we also compared our implementation with theirs numerically. At least for the baselines (e.g. zero-velocity), the numbers were the same up to machine precision.

**Comment by Tim-Salzmann** [comment]
Hi, thanks for the quick reply back then.

Another quick question. I have seen that you are computing the Angle Metric on zyx-Euler Angles.
https://github.com/facebookresearch/QuaterNet/blob/9d8485b732b0a44b99b6cf4b12d3915703507ddc/short_term/pose_network_short_term.py#L67

However, as far as I understand Martinez et al. / RNNexp use xyz-Euler angles:

https://github.com/asheshjain399/RNNexp/blob/3ba986b0032b3c27cae0e221f1e002df7eb8a1ef/structural_rnn/CRFProblems/H3.6m/mhmublv/Motion/RotMat2Euler.m#L26

which is implemented in python by Martinez here:
https://github.com/una-dinosauria/human-motion-prediction/blob/b8f0cb57846e9d2d5688a3e30ab7d8840bb65f38/src/data_utils.py#L16

Using xyz-Angles I can match the the Zero-Velocity of Martinez et all exactly. Using zyx the Zero-Velocity is off for some actions (e.g. smoking). Could you clarify which of the two you are reporting in your paper for your results?

Thanks again!

**Comment by dariopavllo** [comment]
The memory layout of Euler angles is always xyz. However, Euler angles can be composed in different ways depending on the order of multiplication of the individual rotation matrices. This is a convention that must be decided in advance. In this case, the adopted convention is zyx, meaning that you first rotate by z, then by y, then by x. Still, the vector is stored as xyz. Note that this is also explicitly stated in the matlab script you linked: https://github.com/asheshjain399/RNNexp/blob/3ba986b0032b3c27cae0e221f1e002df7eb8a1ef/structural_rnn/CRFProblems/H3.6m/mhmublv/Motion/RotMat2Euler.m#L23

**Comment by Tim-Salzmann** [comment]
Thanks for the reply and thanks for the clarification. However, I tried to follow both, yours and Martinez et al, process from exp_map to euler angles:
`exp = torch.tensor([0.5, 0.5, 0.5])`
`q = torch.tensor(expmap_to_quaternion(exp.numpy()))`
`--> tensor([0.9077, 0.2423, 0.2423, 0.2423])`
`qeuler(q, 'zyx')`
`--> tensor([0.6293, 0.3283, 0.6293])`
`qeuler(q, 'xyz')`
`--> tensor([0.3988, 0.5910, 0.3988])`
`mat = expmap2rotmat(exp.numpy())`
`rotmat2euler(mat)`
`--> array([-0.39876211, -0.59098891, -0.39876211])`

All functions are copied without changes. Looks like 'xyz' leads to the same absolute values while 'zyx' does not. Any idea where I am going wrong?

Thanks again!

**Comment by dariopavllo** [comment]
I think it has to do with the sign. I don't exactly remember why (I worked on this a couple years ago), but IIRC the expmap rotations in the preprocessed dataset have their sign flipped, and if you convert them to Euler angles using an xyz convention, you actually get zyx with a different sign. It might be that the original MATLAB scripts adopted the convention of having a negative sign for the expmap rotation, and all conversion functions were made consistent with this.

Indeed, in our preprocessing script we flip the sign when we import the rotations: https://github.com/facebookresearch/QuaterNet/blob/9d8485b732b0a44b99b6cf4b12d3915703507ddc/prepare_data_short_term.py#L75

To double check, try to run your script with `exp = torch.tensor([-0.5, -0.5, -0.5])` and convert it to Euler angles in zyx order. You should get the same result as `rotmat2euler` (same sign as well).

**Comment by Tim-Salzmann** [comment]
You are absolutely right! I missed that part in your preprocessing. Thanks for taking the time. Much appreciated!
